core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockStore.java

```diff
@@ -0,0 +1,395 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.worker.block;
+
+import alluxio.exception.BlockAlreadyExistsException;
+import alluxio.exception.BlockDoesNotExistException;
+import alluxio.exception.ExceptionMessage;
+import alluxio.exception.UfsBlockAccessTokenUnavailableException;
+import alluxio.worker.block.io.BlockReader;
+import alluxio.worker.block.io.BlockWriter;
+import alluxio.worker.block.meta.UnderFileSystemBlockMeta;
+import alluxio.worker.block.options.OpenUfsBlockOptions;
+
+import com.google.common.base.Objects;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.locks.ReentrantLock;
+
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * This class manages the virtual blocks in the UFS for delegated UFS reads/writes.
+ *
+ * The usage pattern:
+ *  acquireAccess(blockMeta, maxConcurrency)
+ *  cleanup(sessionId, blockId)
+ *  releaseAccess(sessionId, blockId)
+ *
+ * If the client is lost before releasing or cleaning up the session, the session cleaner will
+ * clean the data.
+ */
+public final class UnderFileSystemBlockStore {
+  private static final Logger LOG = LoggerFactory.getLogger(UnderFileSystemBlockStore.class);
+
+  /**
+   * This lock protects mBlocks, mSessionIdToBlockIds and mBlockIdToSessionIds. For any read/write
+   * operations to these maps, the lock needs to be acquired. But once you get the block
+   * information from the map (e.g. mBlocks), the lock does not need to be acquired. For example,
+   * the block reader/writer within the BlockInfo can be updated without acquiring this lock.
+   * This is based on the assumption that one session won't open multiple readers/writers on the
+   * same block. If the client do that, the client can see failures but the worker won't crash.
+   */
```

Comment:
yes, if we can assume that the session won't be reused. Without that assumption, the following can happen:

1. Add <sessionId, blockId> to mBlockIdToSessionIds, mSessionIdToBlockIds, mBlocks
2. Remove <sessionId, blockId> from mBlockIdToSessionIds, mSessionIdToBlockIds but not mBlocks yet.
3. Add <sessionId, blockId> to mBlockIdToSessionIds, mSessionIdToBlockIds. exception will throw since mBlocks contains the pair.
4. step 2 completes. the pair is removed from mBlocks.

Now we end up a state where these 3 maps are inconsistent.  This should rarely happens since sessionId should not collide. If we can assume that, we do not even need locks for the other two maps.

I actually had a TODO to remove this lock initially. but i felt that it might not worth the effort since this UFSBlockStore only keeps track of the active UFS blocks. It should be relatively cheap to access this store compared to access the block itself. I pushed a commit that use ConcurrentHashMap just for mBlocks. I slightly prefer to reverting that commit. How do you think?